source/slang/slang-lower-to-ir.cpp

```diff
@@ -6470,7 +6548,52 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         // body appear before the function itself in the list
         // of global values.
         irFunc->moveToEnd();
-        return LoweredValInfo::simple(finishOuterGenerics(subBuilder, irFunc));
+
+        // If this function is defined inside an interface, add a reference to the IRFunc from
+        // the interface's type definition.
+        auto finalVal = finishOuterGenerics(subBuilder, irFunc, outerGeneric);
+        if (auto genericVal = as<IRGeneric>(finalVal))
+        {
+            auto funcType = lowerFuncType(decl);
+            genericVal->setFullType((IRType*)funcType);
+        }
+
+        maybeAssociateToInterfaceType(decl, finalVal);
+
+        return LoweredValInfo::simple(finalVal);
+    }
+
+    void maybeAssociateToInterfaceType(Decl* decl, IRInst* irFuncVal)
+    {
+        auto parent = decl->parentDecl;
+        InterfaceDecl* interfaceDecl = nullptr;
+        while (parent)
+        {
+            interfaceDecl = as<InterfaceDecl>(parent);
+            if (interfaceDecl) break;
+            parent = parent->parentDecl;
+        }
+        if (!interfaceDecl)
+            return;
+        auto loweredVal = context->findLoweredDecl(interfaceDecl);
+        if (!loweredVal)
+        {
+            return;
+        }
+        IRInst* irFuncType = irFuncVal->typeUse.get();
+        auto irInterfaceType = cast<IRInterfaceType>(loweredVal->val);
+        auto key = getInterfaceRequirementKey(decl);
+        for (UInt i = 0; i < irInterfaceType->getOperandCount(); i++)
+        {
+            auto operand = cast<IRInterfaceRequirementEntry>(irInterfaceType->getOperand(i));
+            if (operand->getOperand(0) == key)
+            {
+                operand->setOperand(1, irFuncType);
+                return;
+            }
```

Comment:
This answers my earlier question about how things are getting "wired up."

The first note is that this seems like it can/should use `operand->getOperands()[1]->init(operand, irFuncType)` instead of going through `setOperand` and then having to tweak that to remove preconditions.

The second note is that it seems like this should be written in a "gather" rather than a "scatter" fashion - the code that builds up the requirement entries should be able to pull these types off of the requirements themselves rather than have the lowering of the requirements push things back into the interface.

Reply:
Interface lowering is now all done in `visitInterfaceDecl` without this injecting logic.